File: codeforces/2051/G. Snakes.py
import sys, math
from collections import deque, defaultdict, Counter

MOD       = 10**9 + 7
FASTIO    = True
DEBUG     = False
INF       = float('inf')
NINF      = float('-inf')
YES, NO   = "YES", "NO"
MT        = False

#====================Solution====================
def sol():
    n, q = ivars()
    who = [0]*(q+1)
    is_plus = [False]*(q+1)
    for t in range(1, q+1):
        si, c = input().split()
        si = int(si)
        who[t] = si
        is_plus[t] = (c == '+')

    E = [[0]*(q+1) for _ in range(n)]
    S = [[0]*(q+1) for _ in range(n)]

    for t in range(1, q+1):
        for i in range(n):
            E[i][t] = E[i][t-1]
            S[i][t] = S[i][t-1]
        snake_id = who[t] - 1
        if is_plus[t]:
            E[snake_id][t] = E[snake_id][t-1] + 1
        else:
            S[snake_id][t] = S[snake_id][t-1] + 1

    MxE = [max(E[i]) for i in range(n)]

    D = [[0]*n for _ in range(n)]

    for i in range(n):
        Ei = E[i]
        for j in range(n):
            Sj = S[j]
            best = -10**15
            for t in range(q+1):
                diff = Ei[t] - Sj[t]
                if diff > best:
                    best = diff
            D[i][j] = best + 1

    INF = 10**18
    dp_maxC = [[INF]*n for _ in range(1<<n)]
    dp_offset = [[0]*n for _ in range(1<<n)]

    for i in range(n):
        mask = (1 << i)
        dp_maxC[mask][i] = 1 + MxE[i]
        dp_offset[mask][i] = 0

    full_mask = (1 << n) - 1

    for mask in range(1<<n):
        m_val = dp_maxC[mask]
        o_val = dp_offset[mask]

        for last in range(n):
            if not (mask & (1<<last)):
                continue
            currMaxC = m_val[last]
            if currMaxC == INF:
                continue
            currOffset = o_val[last]

            others = full_mask ^ mask
            s = others
            while s:
                lowbit = s & -s
                j = (lowbit).bit_length() - 1

                s ^= lowbit

                dist = D[last][j]
                newOffset = currOffset + dist
                newMaxC = max(currMaxC, 1 + newOffset + MxE[j])

                newMask = mask | (1<<j)
                if newMaxC < dp_maxC[newMask][j]:
                    dp_maxC[newMask][j] = newMaxC
                    dp_offset[newMask][j] = newOffset
                elif newMaxC == dp_maxC[newMask][j]:
                    if newOffset < dp_offset[newMask][j]:
                        dp_offset[newMask][j] = newOffset

    ans = min(dp_maxC[full_mask])
    print(ans)


#================================================

def main():
    for _ in range(ipt() if MT else 1):
        sol()


def debug(*args):
    if DEBUG:
        for arg in args:
            print(arg)

def ASSERT(cond):
    if DEBUG:
        assert(cond)


if FASTIO:
    import os
    from io import BytesIO, IOBase

    _str = str
    str = lambda x=b"": x if type(x) is bytes else _str(x).encode()

    BUFSIZE = 8192


    class FastIO(IOBase):
        newlines = 0

        def __init__(self, file):
            self._file = file
            self._fd = file.fileno()
            self.buffer = BytesIO()
            self.writable = "x" in file.mode or "r" not in file.mode
            self.write = self.buffer.write if self.writable else None

        def read(self):
            while True:
                b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
                if not b:
                    break
                ptr = self.buffer.tell()
                self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
            self.newlines = 0
            return self.buffer.read()

        def readline(self):
            while self.newlines == 0:
                b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
                self.newlines = b.count(b"\n") + (not b)
                ptr = self.buffer.tell()
                self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
            self.newlines -= 1
            return self.buffer.readline()

        def flush(self):
            if self.writable:
                os.write(self._fd, self.buffer.getvalue())
                self.buffer.truncate(0), self.buffer.seek(0)


    class IOWrapper(IOBase):
        def __init__(self, file):
            self.buffer = FastIO(file)
            self.flush = self.buffer.flush
            self.writable = self.buffer.writable
            self.write = lambda s: self.buffer.write(s.encode("ascii"))
            self.read = lambda: self.buffer.read().decode("ascii")
            self.readline = lambda: self.buffer.readline().decode("ascii")


    sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
    input = lambda: sys.stdin.readline().rstrip("\r\n")

if __name__ == "__main__":
    finp        = sys.stdin.readline
    njoin       = lambda x: '\n'.join(map(str, x))
    ljoin       = lambda x: ' '.join(map(str, x))
    sjoin       = lambda x: '\n'.join(x)
    ilist       = lambda: list(map(int, finp().split()))
    slist       = lambda: list(input())
    slists      = lambda: input().split()
    ivars       = lambda: map(int, finp().split())
    ipt         = lambda: int(finp())
    main()

File: codeforces/2051/G. Snakes.cpp
#include <bits/stdc++.h>

using namespace std;

#define i64 long long
#define i32 int
#define u32 unsigned int
#define u64 unsigned long long

#define len(x) ((int)x.size())
#define all(a) (a).begin(), (a).end()

const i64 MOD  = 1e9 + 7;
const i64 INF  = 1e9;
const i64 NINF = -(1e9);
const i32 MT   = 1;

const char nl  = '\n';

#include <bits/stdc++.h>
using namespace std;

static const int MAXN = 20;
static const int MAXQ = 200000;


int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;


    static int E[MAXN+1][MAXQ+1];
    static int S[MAXN+1][MAXQ+1];
    for(int i=1; i<=n; i++){
        for(int t=0; t<=q; t++){
            E[i][t] = 0;
            S[i][t] = 0;
        }
    }

    vector<int> who(q+1);
    vector<bool> isPlus(q+1);

    for(int t=1; t<=q; t++){
        int si; char c;
        cin >> si >> c;
        who[t] = si;
        isPlus[t] = (c == '+');
    }

    for(int t=1; t<=q; t++){
        for(int i=1; i<=n; i++){
            E[i][t] = E[i][t-1];
            S[i][t] = S[i][t-1];
        }
        int snakeId = who[t];
        if(isPlus[t]) {
            E[snakeId][t] = E[snakeId][t-1] + 1;
            S[snakeId][t] = S[snakeId][t-1];
        } else {
            S[snakeId][t] = S[snakeId][t-1] + 1;
            E[snakeId][t] = E[snakeId][t-1];
        }
    }

    vector<int> MxE(n+1, 0);
    for(int i=1; i<=n; i++){
        int mx = 0;
        for(int t=0; t<=q; t++){
            mx = max(mx, E[i][t]);
        }
        MxE[i] = mx;
    }

    static int D[MAXN+1][MAXN+1];
    for(int i=1; i<=n; i++){
        for(int j=1; j<=n; j++){
            if(i == j){
                D[i][j] = 1 + 0;
            } else {
                int val = INT_MIN;
                for(int t=0; t<=q; t++){
                    val = max(val, E[i][t] - S[j][t]);
                }
                D[i][j] = val + 1;
            }
        }
    }

    vector<vector<long long>> dp_maxC(1<<n, vector<long long>(n, LLONG_MAX));
    vector<vector<int>> dp_offset(1<<n, vector<int>(n, INT_MAX));

    for(int i=0; i<n; i++){
        int mask = (1 << i);
        dp_offset[mask][i] = 0;
        dp_maxC[mask][i] = 1LL + (long long)(MxE[i+1]);
    }

    for(int mask=0; mask < (1<<n); mask++){
        for(int prev=0; prev<n; prev++){
            if(!(mask & (1<<prev))) continue;
            long long currMaxC = dp_maxC[mask][prev];
            int currOffset = dp_offset[mask][prev];
            if(currMaxC == LLONG_MAX) continue;

            int others = ((1<<n) -1) ^ mask;
            while(others){
                int j = __builtin_ctz(others);
                others ^= (1<<j);

                int dist = D[prev+1][j+1];
                long long newOffset = (long long)currOffset + dist;
                long long newMaxC = max(currMaxC, 1 + newOffset + (long long)MxE[j+1]);

                int newMask = mask | (1<<j);
                if(newMaxC < dp_maxC[newMask][j]){
                    dp_maxC[newMask][j] = newMaxC;
                    dp_offset[newMask][j] = (int)newOffset;
                }
                else if(newMaxC == dp_maxC[newMask][j]){
                    dp_offset[newMask][j] = min(dp_offset[newMask][j], (int)newOffset);
                }
            }
        }
    }

    long long ans = LLONG_MAX;
    int fullMask = (1<<n) - 1;
    for(int last=0; last<n; last++){
        ans = min(ans, dp_maxC[fullMask][last]);
    }

    cout << ans << endl;
    return 0;
}
